### PR TITLE
Compatible with nginx.ingress.kubernetes.io/canary-by-header-pattern annotation

### PR DESCRIPTION
**Pull request by cr7258**
### Ⅰ. Describe what this PR did
Compatible with `nginx.ingress.kubernetes.io/canary-by-header-pattern` annotation.

### Ⅱ. Does this pull request fix one issue?

fixes https://github.com/alibaba/higress/issues/671

### Ⅲ. Test
Manifests
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: production
  labels:
    app: production
spec:
  replicas: 1
  selector:
    matchLabels:
      app: production
  template:
    metadata:
      labels:
        app: production
    spec:
      containers:
      - name: production
        image: registry.k8s.io/ingress-nginx/e2e-test-echo@sha256:6fc5aa2994c86575975bb20a5203651207029a0d28e3f491d8a127d08baadab4
        ports:
        - containerPort: 80
        env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
          - name: POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: POD_IP
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
---
apiVersion: v1
kind: Service
metadata:
  name: production
  labels:
    app: production
spec:
  ports:
  - port: 80
    targetPort: 80
    protocol: TCP
    name: http
  selector:
    app: production
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: canary
  labels:
    app: canary
spec:
  replicas: 1
  selector:
    matchLabels:
      app: canary
  template:
    metadata:
      labels:
        app: canary
    spec:
      containers:
      - name: canary
        image: registry.k8s.io/ingress-nginx/e2e-test-echo@sha256:6fc5aa2994c86575975bb20a5203651207029a0d28e3f491d8a127d08baadab4
        ports:
        - containerPort: 80
        env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
          - name: POD_NAME
            valueFrom:
              fieldRef:
                fieldPath: metadata.name
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: POD_IP
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
---
apiVersion: v1
kind: Service
metadata:
  name: canary
  labels:
    app: canary
spec:
  ports:
  - port: 80
    targetPort: 80
    protocol: TCP
    name: http
  selector:
    app: canary
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: production
spec:
  ingressClassName: higress
  rules:
  - host: echo.prod.mydomain.com
    http:
      paths:
      - pathType: Prefix
        path: /
        backend:
          service:
            name: production
            port:
              number: 80
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: canary
  annotations:
    nginx.ingress.kubernetes.io/canary: "true"
    nginx.ingress.kubernetes.io/canary-by-header: "canary-header"
    nginx.ingress.kubernetes.io/canary-by-header-pattern: "test.XXX.com"
spec:
  ingressClassName: higress
  rules:
  - host: echo.prod.mydomain.com
    http:
      paths:
      - pathType: Prefix
        path: /
        backend:
          service:
            name: canary
            port:
              number: 80
```

Test request.
```bash
kubectl port-forward  -n higress-system  svc/higress-gateway  18080:80

for i in $(seq 1 10); do curl -s --resolve echo.prod.mydomain.com:18080:127.0.0.1 echo.prod.mydomain.com:18080 \
-H "canary-header: http://test.XXX.com/YYY" | grep "Hostname"; done

Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd


for i in $(seq 1 10); do curl -s --resolve echo.prod.mydomain.com:18080:127.0.0.1 echo.prod.mydomain.com:18080 \
-H "canary-header: http://test.XXX.com/ZZZ" | grep "Hostname"; done

Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd
Hostname: canary-847fbf5f84-9wprd


for i in $(seq 1 10); do curl -s --resolve echo.prod.mydomain.com:18080:127.0.0.1 echo.prod.mydomain.com:18080 \
-H "canary-header: http://test.YYY.com" | grep "Hostname"; done

Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
Hostname: production-b9c8f68b6-j6cll
```

### CODE REVIEW

**Comment by johnlanni** [comment]
@cr7258 Please add some UT or e2e test for it.

**Comment by cr7258** [comment]
@johnlanni e2e test added.
```
 --- PASS: TestHigressConformanceTests/HTTPRouteCanaryHeader (1.78s)
        --- PASS: TestHigressConformanceTests/HTTPRouteCanaryHeader/Canary_HTTPRoute_Traffic_Split (1.66s)
```

File: pkg/ingress/kube/annotations/canary.go [hunk 1]
```diff
@@ -154,7 +154,7 @@ func ApplyByHeader(canary, route *networking.HTTPRoute, canaryIngress *Ingress)
 				match.Headers = map[string]*networking.StringMatch{
 					canaryConfig.Header: {
 						MatchType: &networking.StringMatch_Regex{
-							Regex: canaryConfig.HeaderPattern,
+							Regex: ".*" + canaryConfig.HeaderPattern + ".*",
 						},
```
Comment (SpecialYang) [inline pkg/ingress/kube/annotations/canary.go:158]: 如果用户的正则表达式是 ^test$，那么行为是什么样子。是不是也是只匹配test，而非xxtestxx

Comment (cr7258) [reply pkg/ingress/kube/annotations/canary.go:158]: 是的，只会只会匹配test。

Comment (SpecialYang) [reply pkg/ingress/kube/annotations/canary.go:158]: 可以加个这种情况的单测吗？

**Review by SpecialYang — APPROVED** [review]
Thanks.

**Pull request merged**